aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -342,18 +343,49 @@ Tensor& eye_out_cpu(Tensor& result, int64_t n, int64_t m) {
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ full ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Tensor full(IntArrayRef size, Scalar fill_value, const TensorOptions& options) {
-  if (options.layout() == kSparse) {
-    AT_ERROR("full(...) is not implemented for sparse layout");
+namespace {
+
+// Performs dtype inference for full
+TensorOptions infer_full_options(
+  Scalar fill_value,
+  const TensorOptions& options) {
+
+  if (!options.has_dtype()) {
+    if (fill_value.isIntegral(true)) {
+      TORCH_WARN_ONCE(
+        "Deprecation warning: In PyTorch 1.6 torch.full with a bool or ",
+        "integral fill_value will require a dtype or out argument. ",
+        "In PyTorch 1.7, when `out` and `dtype` are not set a bool fill_value ",
+        "will return a tensor of torch.bool dtype, ",
```

Comment:
I think we should give separate messages for if the user called the out variant or not.

The reasoning is basically, I don't think the out variants are a particularly good idea -- because these are factory functions, they specify properties of the memory of the tensor, i.e. the memory format.  What does it mean for the memory format to be one thing, but the out argument to be another?  Best if we limit the usage of the `out=` variants.

Reply:
The error message for setting the out= kwarg improperly is in python_torch_functions.py. I added a test for the behavior.